### PR TITLE
Add feature flag to opt in for kctrl pkg cmd tree

### LINKED ISSUE #1865: Package plugin and kctrl integration
(This is used to request new product features)

**Describe the feature request**
Use [kctrl](https://carvel.dev/blog/kctrl-release-blog/) to add a command tree to `tanzu package plugin`. This would help in having a consistent experience for Carvel users.

**Describe alternatives you've considered**

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [x] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [x] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by praveenrewar**
### What this PR does / why we need it
- [x] Add package command tree from kctrl to package plugin
- [x] Use `tanzu config set features.package.kctrl-package-command-tree true` to enable the kctrl flow.
- [x] Default value for `features.package.kctrl-package-command-tree` is false.

### Which issue(s) this PR fixes

Fixes #1865

### Describe testing done for PR

- [x] tkg package plugin integration tests
- [x] Manual testing on workload cluster

### Release note

```release-note
- package plugin now provides extended functionality for better observability and debugging.
  - `tanzu config set features.package.kctrl-package-command-tree true` can be used to enable this new behaviour.
  - package installed includes status command which can be used to tail status of a package install.
  - package installed includes pause and kick commands which can be used to pause the reconciliation of a package install or force trigger the reconciliation of a package install.
  - The tailing behaviour for a package install is also present in kick command and has been added to package installed create or update.
  - Breaking changes when feature flag is activated:
    - `--verbose` global flag is removed.
    - `--create-namespace` flag is removed from all the commands (`package repository add`, `package repository update`, `package installed create`, `package installed update` and `package install`)
    - `--poll-interval` and `-poll-timeout` are renamed to `--wait-interval` and `--wait-timeout` in all the commands (`package repository add`, `package repository update`, `package repository delete`, `package install`, `package installed create` and `package installed update`.
    - `--create` flag is removed in `package repository update`.
    - REPOSITORY and TAG columns are replaced by `SOURCE` column which consist of source type (eg imgpkg), repo url and tag in `package repository list`.
    - In `package available list`, SHORT-DESCRIPTION is not displayed by default (--wide flag can be used to display it). Also, LATEST-VERSION column is removed.
    - `package available get` provides an additional table which lists the available versions for that package.
    - `--generate-default-values-file` is renamed to `--default-values-file-output` in `package available get` which accepts a filename where the default values need to be saved. Also, all the values are commented by default.
    - `--package-name` flag is renamed to `--package` in `package installed create` and `package install`.
    - `--install` flag is removed in the `package installed update`.

```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: cmd/cli/plugin/package/kctrl/README.md [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+  tanzu package [flags]
+  tanzu package [command]
+
+Available Commands:
+  available     Manage available packages (get, list)
+  completion    Generate the autocompletion script for the specified shell
+  install       Install package
+  installed     Manage installed packages (create, delete, get, list, update)
+  repository    Manage package repositories (add, delete, get, list, update)
+
```
Comment (maralavi) [inline cmd/cli/plugin/package/kctrl/README.md:23]: Wondering if the Readme content should be replaced or not considering that kctrl package plugin is gonna be behind feature flag for now.

File: cmd/cli/plugin/package/main.go [hunk 2]
```diff
@@ -33,6 +35,13 @@ func main() {
 		log.Fatal(err)
 	}
 
+	if config.IsFeatureActivated(config.FeatureFlagPackagePluginKctrlCommandTree) {
+		if err := kctrl.Invoke(p); err != nil {
+			os.Exit(1)
+		}
+		return
```
Comment (maralavi) [inline cmd/cli/plugin/package/main.go:42]: Could I ask you please test this manually by disabling this feature flag and making sure legacy package plugin commands are working properly?

File: cmd/cli/plugin/package/kctrl/README.md (hunk 1 above)
Comment (praveenrewar) [reply cmd/cli/plugin/package/kctrl/README.md:23]: > kctrl package plugin is gonna be behind feature flag for now.

Yep, considering that, I have not updated the existing README, I have created a new one in the kctrl package. We can replace the existing README with the new one when we decide to make this as the default behaviour.

File: cmd/cli/plugin/package/main.go (hunk 2 above)
Comment (praveenrewar) [reply cmd/cli/plugin/package/main.go:42]: Yep, I have tested with both the feature flag enabled and disabled. I have also not updated the existing package plugin integration tests so they are also passing for the existing behaviour of package plugin (i.e. the feature flag being disabled).

File: pkg/v1/config/clientconfig.go [hunk 3]
```diff
@@ -57,6 +57,9 @@ const (
 	// TKR version v1alpha3 feature flag determines whether to use Tanzu Kubernetes Release API version v1alpha3. Setting
 	// feature flag to true will allow to use the TKR version v1alpha3; false allows to use legacy TKR version v1alpha1
 	FeatureFlagTKRVersionV1Alpha3 = "features.global.tkr-version-v1alpha3-beta"
+	// Package Plugin Kctrl Command Tree determines whether to use the command tree from kctrl. Setting feature flag to
+	// true will allow to use the package command tree from kctrl for package plugin
+	FeatureFlagPackagePluginKctrlCommandTree = "features.package.kctrl-package-command-tree"
```
Comment (praveenrewar) [inline pkg/v1/config/clientconfig.go:62]: Is there any convention on deciding the feature flag name? I am open to suggestions on this.

File: cmd/cli/plugin/package/kctrl/ui.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	}
+}
+
+func (adapterUI *AdapterUI) SetOutputFormat(outputFormat string) {
+	adapterUI.outputFormat = outputFormat
+}
+
+// PrintTable overrides go-cli-ui/ui.PrintTable
+// It accepts a table and renders it based on the output format
+//nolint:gocritic // Cannot change the function signature as it is defined in go-cli-ui
```
Comment (vijaykatam) [inline cmd/cli/plugin/package/kctrl/ui.go:46]: Not blocking PR for this but as I mentioned in previous discussion, use a different func signature.

**Review by vijaykatam — APPROVED** [review]

**Comment by renuy** [comment]
@vijaykatam , who will merge these changes? can @praveenrewar go ahead and merge these? when can this be merged?

**Pull request merged**